api/fixtures_test.go

```diff
@@ -0,0 +1,528 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/ledger"
+
+	"github.com/algorand/indexer/api/generated/v2"
+	"github.com/algorand/indexer/idb/postgres"
+	"github.com/algorand/indexer/processor"
+	"github.com/algorand/indexer/util/test"
+)
+
+/* **************************************
+This Fixtures Test simulates calling indexer GET endpoints and in so doing validates:
+* Hand crafted URL's are:
+	* routed to the correct endpoint
+	* handled as expected
+* HTTP Responses are:
+	* well formatted and can be parsed to produce results of the expected type
+	* have the state that is expected for a "real" blockchain
+
+This is achieved as follows:
+A. Set up the state of the blockchain
+B. Iterate through a hard-coded `xyzSeedFixture` to generate an `xyzLiveFixture` as follows. For each seed object in `xyzSeedFixture`:
+	1. Query indexer via HTTP for the state of the blockchain using the `seed`'s `Request.Path` + `Request.Params`
+	2. Query the test-internal `proverRoutes` struct for a route-compatible `prover`
+	3. Employ the `prover` to parse the indexer's response into a route appropriate `witness` object
+C. Save `xyzLiveFixture` into a non-git-committed fixture `./test_resources/_FIXTURE_NAME.json` (NOTICE THE `_` PREFIX)
+D. Assert that the non-git-commited fixture equals the git-committed version `./test_resources/FIXTURE_NAME.json`
+
+It is the responsibility of the test writer to check that the generated fixture represent the expected results.
+
+----
+TODO: SHOULD this be used to auto-generate unit test cases in the SDK's?
+In particular, one could craft a generic cucumber test that looks something like:
```

Comment:
Any thoughts?
@winder @shiqizng @Eric-Warehime @AlgoStephenAkiki @algochoi @michaeldiamant

Reply:
I've used [httpmock](https://github.com/algorand/indexer/blob/develop/processor/blockprocessor/initialize_test.go#L33). I think it's serving similar purpose to the fixtures you're writing.

Reply:
Thanks for the tip. I'll research that tool a bit to see if I actually re-invented the wheel and what the pros/cons were.

Reply:
@tzaffi I think I'd benefit from a live walkthrough to better understand the tradeoffs considered here.

I've spent ~20min reviewing the space to build _some_ awareness, though I suspect I don't see the entire space.

Reply:
Leaving a note after having a verbal with @tzaffi + @shiqizng with my 2c:

* I see overlap between `fixtures_test.go` and `handlers_e2e_test.go`.  I think a primary question for Indexer maintainers to consider is:  What invariants ought to be maintained when modifying + adding endpoints? 
* I'd consider comfort levels with adopting a `fixture_test.go` approach _in place of_ `handlers_e2e_test.go` rather than trying to maintain both approaches.
  * From my perspective, `fixtures_test.go` brings more rigor to API-level testing.  I can see the approach making it easier to catch API regressions in exchange for maintaining a different approach to API-level tests.
  * Ultimately, I think indexer maintainers ought to weigh in on the tradeoff they feel is appropriate.  If helpful, I'm happy to further vocalize my opinion, but I don't think I'm in a decision making role here.
* RE: Usage for SDK test generation - My initial reaction is that the upfront investment is too large relative to the perceived payoff.  So, I'd focus evaluation on usage within indexer without considering downstream SDK testing possibilities.

Reply:
Took a brief look, and I agree...the fixtures tests both give me more confidence in the results, and provide more detail on expected vs actual. Some of the existing API handlers tests are incredibly difficult to manually parse/debug because of a combination of b64/msgp encoded data.

I also agree that we should not maintain both sets of tests. 

If the team agrees, let's do the following:
* For the short term I think it should be fine to add these and require that any new API handler tests in the future be added as fixtures. 
* In the medium/longer term let's open an issue for converting the existing tests to the new fixture format.

Reply:
Not sure how the fixtures test would handle msgp. It's worth further investigation.